packages/eslint-plugin/src/rules/parameter-properties.ts

```diff
@@ -0,0 +1,239 @@
+
+    function typeAnnotationsMatch(
+      classProperty: TSESTree.PropertyDefinition,
+      constructorParameter: TSESTree.Identifier,
+    ): boolean {
+      if (
+        !classProperty.typeAnnotation ||
+        !constructorParameter.typeAnnotation
+      ) {
+        return (
+          classProperty.typeAnnotation === constructorParameter.typeAnnotation
+        );
+      }
+
+      return (
+        sourceCode.getText(classProperty.typeAnnotation) ===
+        sourceCode.getText(constructorParameter.typeAnnotation)
+      );
+    }
+
+    return {
+      'ClassDeclaration, ClassExpression'(): void {
+        propertyNodesByNameStack.push(new Map());
+      },
+
+      ':matches(ClassDeclaration, ClassExpression):exit'(): void {
+        const propertyNodesByName = propertyNodesByNameStack.pop()!;
+
+        for (const [name, nodes] of propertyNodesByName) {
+          if (
+            nodes.classProperty &&
+            nodes.constructorAssignment &&
+            nodes.constructorParameter &&
+            typeAnnotationsMatch(
+              nodes.classProperty,
+              nodes.constructorParameter,
+            )
+          ) {
+            context.report({
+              data: {
+                parameter: name,
+              },
+              messageId: 'preferParameterProperty',
+              node: nodes.classProperty,
+            });
+          }
+        }
+      },
+
+      ClassBody(node): void {
+        for (const element of node.body) {
+          if (
+            element.type === AST_NODE_TYPES.PropertyDefinition &&
+            element.key.type === AST_NODE_TYPES.Identifier &&
+            !element.value &&
+            !allow.includes(getModifiers(element))
+          ) {
+            getNodesByName(element.key.name).classProperty = element;
+          }
+        }
+      },
+
+      'MethodDefinition[kind="constructor"]'(
+        node: TSESTree.MethodDefinition,
+      ): void {
+        for (const parameter of node.value.params) {
+          if (parameter.type === AST_NODE_TYPES.Identifier) {
+            getNodesByName(parameter.name).constructorParameter = parameter;
+          }
+        }
+
+        for (const statement of node.value.body?.body ?? []) {
+          if (
+            statement.type !== AST_NODE_TYPES.ExpressionStatement ||
+            statement.expression.type !== AST_NODE_TYPES.AssignmentExpression ||
+            statement.expression.left.type !==
+              AST_NODE_TYPES.MemberExpression ||
+            statement.expression.left.object.type !==
+              AST_NODE_TYPES.ThisExpression ||
+            statement.expression.left.property.type !==
+              AST_NODE_TYPES.Identifier ||
+            statement.expression.right.type !== AST_NODE_TYPES.Identifier
+          ) {
+            break;
+          }
+
+          getNodesByName(
+            statement.expression.right.name,
+          ).constructorAssignment = statement.expression;
+        }
```

Comment:
This is a good start, but it won't handle cases that aren't top-level:
```ts
constructor(foo) {
  {
    this.foo = foo;
  }
}
```

I'm not sure if we need to handle these cases?
If we need to - we'll need to refactor this a bit to leverage the same stack pattern as above, like:
```ts
let currentConstructor = null;
return {
  'MethodDefinition[kind="constructor"]'(node) {
    currentConstructor = { parent: currentConstructor, node };
    // ...
  },
  'MethodDefinition[kind="constructor"]:exit'(node) {
    currentConstructor = currentConstructor?.parent;
  },
  'AssignmentExpression > MemberExpression.left > ThisExpression.object'(node) {
    // this.<...> = <...>
    if (currentConstructor == null) {
      return;
    }
    // ...
  },
}
```

Reply:
I don't think we want to handle those cases, no. TypeScript's parameter properties are equivalent only to placing the `this.... = ...` statements at the very beginning of the constructor. It's pretty rare that user code would have anything like the block snippet/